FLiMESolve

magnamancer opened the pull request:
**Description**
Added FLiMESolve. Changed head to master branch after merging QuTiP 5.0.X to master. 
Old PR here: https://github.com/qutip/qutip/pull/2140

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,705 @@
+        Default value of zero uses the fully time-independent/most strict
+        secular approximation.
+
+    quicksolve: Boolean
```
The function does not seems to use it...

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,705 @@
+            state = state.copy()
+        return state
+
+    def run(self, state00, taulist, *, floquet=False, args=None, e_ops=None,):
```
`state00`?

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,705 @@
+
+        Parameters
+        ----------
+        state0 : :class:`Qobj`
```
`state0` or `state00`, they should at least match.

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,705 @@
+            tally = {}
+            for i, item in enumerate(taulist_test_new):
+                try:
+                    tally[item].append(
+                        i)
```
You have a lot of this strange line splitting...
When it fit on one line, it should not be split.

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,705 @@
+        ) for RateMat in RateDic.values()]
+        self.R0 = Rate_Qobj_list[0]
+
+        self.Rt_timedep_pairs = []
```
These `self.Rt_timedep_pairs` are never used?

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,705 @@
+        if args:
+            raise ValueError(
+                "FMESolver cannot update arguments")
+        state = super().step(t)
```
Did you test this?
I am pretty sure it would not work when `time_sense=0`.

Ericgig replied:
This was fixed a while back.

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,705 @@
+                                                   'exp(1j*' + str(
+                    key * list(Hargs.values())[0])
+                    + '*t)']))
+        self.Rt_timedep_pairs = [list([Rate_Qobj_list[idx],
```
This is defined twice...

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,705 @@
+    rho0 / psi0 : :class:`qutip.Qobj`
+        Initial density matrix or state vector (ket).
+
+    Taulist:*list* / *array*
```
Since time dependent c_ops not being supported is the exception, it would be nice to have a note in this regard.

magnamancer replied:
I'll make a comment with this, but I _think_ time-dependent collapse c_ops would be implementable. I just wasn't sure how to do it when I was writing this up. I may try to write them in as I refactor the physics and data operations to be separate.

Ericgig reviewed (changes requested):
Are you sure there was not a bad merge somewhere?
Some code is duplicated, seemingly important variable are never used, a typo in the import of flimesolve.py stop it from being imported, etc...

Please add tests covering most case and have them passing.

Also the separation of task is not well used. We split the data container, ODE integration and the physics, but here they are mixed together...

The quicksolve is the "diag" integrator, you don't have to re implement it here. You could just set it as the default method since the rhs is always constant (but it seems like this is not desired when time_sense != 0).

magnamancer commented:
I'll take a look at the merge. It seems like it might have gone bad.

I'll add in some tests once I fix the merge, as well.

I'll look into the diag method and try to move everything over to separate
everything as you described.

On Wed, Jun 28, 2023, 3:43 PM Eric Giguère ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> Are you sure there was not a bad merge somewhere?
> Some code is duplicated, seemingly important variable are never used, a
> typo in the import of flimesolve.py stop it from being imported, etc...
>
> Please add tests covering most case and have them passing.
>
> Also the separation of task is not well used. We split the data container,
> ODE integration and the physics, but here they are mixed together...
>
> The quicksolve is the "diag" integrator, you don't have to re implement it
> here. You could just set it as the default method since the rhs is always
> constant (but it seems like this is not desired when time_sense != 0).
> ------------------------------
>
> In doc/guide/scripts/floquet_ex3.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245104940>:
>
> > @@ -35,7 +35,6 @@ def noise_spectrum(omega):
>      # Alternatively
>      psi_t = output.states[idx]
>      p_ex[idx] = qutip.expect(qutip.num(2), psi_t)
> -
>
> Why did you erase this line?
> ------------------------------
>
> In VERSION
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245585117>:
>
> > @@ -1 +1 @@
> -5.0.0.dev
> +5.0.0a1
>
> The version change when we do a release, not with each PR.
> ------------------------------
>
> In qutip/solver/correlation.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245591365>:
>
> > +    G1 = correlation_3op(
> +        solver, state0, [0], taulist, None, a_op.dag(), a_op)[0]
>
> ⬇️ Suggested change
>
> -    G1 = correlation_3op(
> -        solver, state0, [0], taulist, None, a_op.dag(), a_op)[0]
> +    G1 = correlation_3op(
> +        solver, state0, [0], taulist, None, a_op.dag(), a_op
> +    )[0]
>
> ------------------------------
>
> In qutip/solver/correlation.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245592358>:
>
> > +        solver_instance = FLiMESolver(
> +            floquet_basis,
> +            c_ops,
> +            args,
> +            time_sense=time_sense
> +        )
>
> ⬇️ Suggested change
>
> -        solver_instance = FLiMESolver(
> -            floquet_basis,
> -            c_ops,
> -            args,
> -            time_sense=time_sense
> -        )
> +        solver_instance = FLiMESolver(
> +            floquet_basis,
> +            c_ops,
> +            args,
> +            time_sense=time_sense,
> +            options=options,
> +        )
>
> ------------------------------
>
> In qutip/solver/floquet.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245594985>:
>
> > @@ -929,4 +930,5 @@ def run(self, state0, tlist, *, floquet=False, args=None, e_ops=None):
>          stats["run time"] = progress_bar.total_time()
>          # TODO: It would be nice if integrator could give evolution statistics
>          # stats.update(_integrator.stats)
> +
>
> Why a new empty line here?
> ------------------------------
>
> In qutip/solver/floquet.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245604228>:
>
> > +            self.evecs = _data.permute.indices(
> +                evecs, col_perm=np.argsort(perm))
>
> This file already has black ran on it and the style is ok as is.
> (But I just learned that black exclude the newline in it's character count
> while pycodestyle include it.)
> Since it's not the main part of your PR, please revert it to it's original
> state.
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245610200>:
>
> > +                total_R_tensor[key] += np.reshape(flime_FirstTerm - (1 / 2)
> +                                                  * (flime_SecondTerm +
> +                                                     flime_ThirdTerm),
> +                                                  (Hdim**2, Hdim**2))
> +            except KeyError:
> +                total_R_tensor[key] = np.reshape(flime_FirstTerm - (1 / 2)
> +                                                 * (flime_SecondTerm +
> +                                                    flime_ThirdTerm),
> +                                                 (Hdim**2, Hdim**2))
> +    return total_R_tensor
> +
> +
> +def flimesolve(
> +        H,
> +        rho0,
> +        taulist,
>
> ⬇️ Suggested change
>
> -        taulist,
> +        tlist,
>
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245610703>:
>
> > +        time_sense=0,
> +        quicksolve=False,
> +        options=None):
> +    """
> +    Parameters
> +    ----------
> +
> +    H : :class:`Qobj`,:class:`QobjEvo`,:class:`QobjEvo` compatible format.
> +        Periodic system Hamiltonian as :class:`QobjEvo`. List of
> +        [:class:`Qobj`,:class:`Coefficient`] or callable that
> +        can be made into :class:`QobjEvo` are also accepted.
> +
> +    rho0 / psi0 : :class:`qutip.Qobj`
> +        Initial density matrix or state vector (ket).
> +
> +    Taulist:*list* / *array*
>
> ⬇️ Suggested change
>
> -    Taulist:*list* / *array*
> +    tlist: *list* / *array*
>
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245610938>:
>
> > +    H : :class:`Qobj`,:class:`QobjEvo`,:class:`QobjEvo` compatible format.
> +        Periodic system Hamiltonian as :class:`QobjEvo`. List of
> +        [:class:`Qobj`,:class:`Coefficient`] or callable that
> +        can be made into :class:`QobjEvo` are also accepted.
> +
> +    rho0 / psi0 : :class:`qutip.Qobj`
> +        Initial density matrix or state vector (ket).
> +
> +    Taulist:*list* / *array*
> +        List of times for :math:`t`.
> +
> +    T : float
> +        The period of the time-dependence of the hamiltonian.
> +
> +    c_ops_and_rates : list of :class:`qutip.Qobj`.
> +        List of lists of [collapse operator,collapse operator rate] pairs
>
> ⬇️ Suggested change
>
> -        List of lists of [collapse operator,collapse operator rate] pairs
> +        List of lists of [collapse operator, collapse operator rate] pairs
>
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245611168>:
>
> > +
> +    e_ops : list of :class:`qutip.Qobj` / callback function
> +        List of operators for which to evaluate expectation values.
> +        The states are reverted to the lab basis before applying the
> +
> +    args : *dictionary*
> +        Dictionary of parameters for time-dependent Hamiltonian
> +
> +    time_sense : float
> +        Experimental. Value of the secular approximation (in terms of system
> +        frequency 2*np.pi/T) to use when constructing the rate matrix R(t).
> +        Default value of zero uses the fully time-independent/most strict
> +        secular approximation.
> +
> +    quicksolve: Boolean
> +        True to use the quicksolve method,which utilizes the most strict
>
> ⬇️ Suggested change
>
> -        True to use the quicksolve method,which utilizes the most strict
> +        True to use the quicksolve method, which utilizes the most strict
>
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245613247>:
>
> > +        IVP solvers. Can be overridden if desired (e.g. for
> +                                                   debuggin/troubleshooting).
>
> With overwrite you mean using quicksolve with non-zero time_sense?
> If so, how to overwrite it?
> ⬇️ Suggested change
>
> -        IVP solvers. Can be overridden if desired (e.g. for
> -                                                   debuggin/troubleshooting).
> +        IVP solvers. Can be overridden if desired (e.g. for
> +        debuggin/troubleshooting).
>
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245614869>:
>
> > +        List of lists of [collapse operator,collapse operator rate] pairs
> +
> +    e_ops : list of :class:`qutip.Qobj` / callback function
> +        List of operators for which to evaluate expectation values.
> +        The states are reverted to the lab basis before applying the
> +
> +    args : *dictionary*
> +        Dictionary of parameters for time-dependent Hamiltonian
> +
> +    time_sense : float
> +        Experimental. Value of the secular approximation (in terms of system
> +        frequency 2*np.pi/T) to use when constructing the rate matrix R(t).
> +        Default value of zero uses the fully time-independent/most strict
> +        secular approximation.
> +
> +    quicksolve: Boolean
>
> Input that change how the computation is done should be in options.
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245616491>:
>
> > +                                                     flime_ThirdTerm),
> +                                                  (Hdim**2, Hdim**2))
> +            except KeyError:
> +                total_R_tensor[key] = np.reshape(flime_FirstTerm - (1 / 2)
> +                                                 * (flime_SecondTerm +
> +                                                    flime_ThirdTerm),
> +                                                 (Hdim**2, Hdim**2))
> +    return total_R_tensor
> +
> +
> +def flimesolve(
> +        H,
> +        rho0,
> +        taulist,
> +        T,
> +        Nt=None,
>
> There is no docstring entry for Nt.
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245620508>:
>
> > +        if isinstance(floquet_basis, FloquetBasis):
> +            self.floquet_basis = floquet_basis
> +        else:
> +            raise TypeError(
> +                "The ``floquet_basis`` must be a FloquetBasis")
>
> Check done twice
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245632475>:
>
> > +        Rate_Qobj_list = [Qobj(
> +            RateMat, dims=[[self.Hdim, self.Hdim], [
> +                self.Hdim, self.Hdim]],
>
> You go the Hdim from the shape, not the dimension. dims has the
> information about subsystem, shape does not. This should make test fail if
> there where some.
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245632794>:
>
> > +        c_ops = []
> +        c_op_rates = []
> +        for entry in c_ops_and_rates:
> +            c_ops.append(entry[0])
> +            c_op_rates.append(entry[1])
>
> Small trick:
> ⬇️ Suggested change
>
> -        c_ops = []
> -        c_op_rates = []
> -        for entry in c_ops_and_rates:
> -            c_ops.append(entry[0])
> -            c_op_rates.append(entry[1])
> +        c_ops, c_op_rates = zip(*c_ops_and_rates)
>
> or do the check here while you are looping over all elements.
> ⬇️ Suggested change
>
> -        c_ops = []
> -        c_op_rates = []
> -        for entry in c_ops_and_rates:
> -            c_ops.append(entry[0])
> -            c_op_rates.append(entry[1])
> +        c_ops = []
> +        c_op_rates = []
> +        for c_op, rate in c_ops_and_rates:
> +            if not isinstance(c_op, Qobj):
> +                raise TypeError("c_ops must be type Qobj")
> +            c_ops.append(c_op)
> +            c_op_rates.append(rate)
>
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245642075>:
>
> > +            The state must be initialized first by calling ``start`` or
> +            ``run``. If ``run`` is called,``step`` will continue from the last
> +            time and state obtained.
> +        """
> +        if args:
> +            raise ValueError(
> +                "FMESolver cannot update arguments")
> +        state = super().step(t)
> +        if not floquet:
> +            state = self.floquet_basis.from_floquet_basis(
> +                state, t)
> +        elif copy:
> +            state = state.copy()
> +        return state
> +
> +    def run(self, state00, taulist, *, floquet=False, args=None, e_ops=None,):
>
> state00?
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245643705>:
>
> > +        elif copy:
> +            state = state.copy()
> +        return state
> +
> +    def run(self, state00, taulist, *, floquet=False, args=None, e_ops=None,):
> +        """
> +        Calculate the evolution of the quantum system.
> +
> +        For a ``state0`` at time ``tlist[0]`` do the evolution as directed by
> +        ``rhs`` and for each time in ``tlist`` store the state and/or
> +        expectation values in a :class:`Result`. The evolution method and
> +        stored results are determined by ``options``.
> +
> +        Parameters
> +        ----------
> +        state0 : :class:`Qobj`
>
> state0 or state00, they should at least match.
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245648476>:
>
> > +                    tally[item].append(
> +                        i)
>
> You have a lot of this strange line splitting...
> When it fit on one line, it should not be split.
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245663589>:
>
> > +            self.floquet_basis.T,
> +            c_ops,
> +            c_op_rates,
> +            Hargs,
> +            time_sense=time_sense)
> +
> +        Rate_Qobj_list = [Qobj(
> +            RateMat, dims=[[self.Hdim, self.Hdim], [
> +                self.Hdim, self.Hdim]],
> +            type="super",
> +            superrep="super",
> +            copy=False
> +        ) for RateMat in RateDic.values()]
> +        self.R0 = Rate_Qobj_list[0]
> +
> +        self.Rt_timedep_pairs = []
>
> These self.Rt_timedep_pairs are never used?
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245667295>:
>
> > +        floquet : bool,optional {False}
> +            Whether to return the state in the floquet basis or laboratory
> +            basis.
> +
> +        args : dict,optional {None}
> +            Not supported
> +
> +        .. note::
> +            The state must be initialized first by calling ``start`` or
> +            ``run``. If ``run`` is called,``step`` will continue from the last
> +            time and state obtained.
> +        """
> +        if args:
> +            raise ValueError(
> +                "FMESolver cannot update arguments")
> +        state = super().step(t)
>
> Did you test this?
> I am pretty sure it would not work when time_sense=0.
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245668542>:
>
> > +            RateMat, dims=[[self.Hdim, self.Hdim], [
> +                self.Hdim, self.Hdim]],
> +            type="super",
> +            superrep="super",
> +            copy=False
> +        ) for RateMat in RateDic.values()]
> +        self.R0 = Rate_Qobj_list[0]
> +
> +        self.Rt_timedep_pairs = []
> +        for idx, key in enumerate(RateDic.keys()):
> +            if key != 0.0:
> +                self.Rt_timedep_pairs.append(list([Rate_Qobj_list[idx],
> +                                                   'exp(1j*' + str(
> +                    key * list(Hargs.values())[0])
> +                    + '*t)']))
> +        self.Rt_timedep_pairs = [list([Rate_Qobj_list[idx],
>
> This is defined twice...
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245675791>:
>
> > +        time_sense=0,
> +        quicksolve=False,
> +        options=None):
> +    """
> +    Parameters
> +    ----------
> +
> +    H : :class:`Qobj`,:class:`QobjEvo`,:class:`QobjEvo` compatible format.
> +        Periodic system Hamiltonian as :class:`QobjEvo`. List of
> +        [:class:`Qobj`,:class:`Coefficient`] or callable that
> +        can be made into :class:`QobjEvo` are also accepted.
> +
> +    rho0 / psi0 : :class:`qutip.Qobj`
> +        Initial density matrix or state vector (ket).
> +
> +    Taulist:*list* / *array*
>
> Since time dependent c_ops not being supported is the exception, it would
> be nice to have a note in this regard.
> ------------------------------
>
> In qutip/solver/flimesolve.py
> <https://github.com/qutip/qutip/pull/2186#discussion_r1245633903>:
>
> > +        List of lists of [collapse operator,collapse operator rate] pairs
> +
> +    e_ops : list of :class:`qutip.Qobj` / callback function
> +        List of operators for which to evaluate expectation values.
> +        The states are reverted to the lab basis before applying the
> +
> +    args : *dictionary*
> +        Dictionary of parameters for time-dependent Hamiltonian
> +
> +    time_sense : float
> +        Experimental. Value of the secular approximation (in terms of system
> +        frequency 2*np.pi/T) to use when constructing the rate matrix R(t).
> +        Default value of zero uses the fully time-independent/most strict
> +        secular approximation.
> +
> +    quicksolve: Boolean
>
> The function does not seems to use it...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/qutip/qutip/pull/2186#pullrequestreview-1502921416>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AV5WL5AGKQRG75CR5YOQUD3XNSCP7ANCNFSM6AAAAAAZWI6GRQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

magnamancer commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,705 @@
+        matrix-exponential method instead of QuTiP internal IVP solvers.
+        Automatically updates based on time_sense,with the default time_sense
+        value 0 using quicksolve,and any other time-sense value using the
+        IVP solvers. Can be overridden if desired (e.g. for
+                                                   debuggin/troubleshooting).
```
I think I may end up writing out time_solve as an input argument altogether. It can be overwritten by making quicksolve = True, which will use the quicksolve method even if a time dependence is desired. However, I feel like this may add confusion and ultimately doesn't provide any extra functionality, such that I don't think it'll be worthwhile to keep this as it is.

hodgestar commented:
@magnamancer I think part of the issue was that you merged `qutip-5.0.X` into `master` but that shouldn't have been done because `master` now contains the latest 5.X qutip code.

Sorry to be a pain, but could you open a new new PR with just your changes? No need to be fancy -- just create a new branch from master, copy across the files you want to change, create PR.

The `VERSION` file should not be changed.

How hard would it be to add tests? I remember you have tests in a notebook somewhere.  If you can point me at the latest versions, I can think about how to convert them into unit tests and add them to the new PR.

magnamancer commented:
Hey Simon,

I've been working on a few issues this last week that Eric pointed out.

I actually did end up going back and just copying the master branch to my
local and copying in my changes, so hopefully that won't be an issue going
forward. I'll likely do it a second time before submitting another request
just to be sure there wasn't anything I changed that I've forgotten about.

I've also been working on adding in a few tests. I'm mostly working on
mirroring the floquet_test test file as best I can, although it looks like
there are some slight changes I need to make that I've been working on.

Beyond that, I've been working on adding a third system to this paper I'm
working on that describes FLiME. This third system should be able to show
changes based on the time_sense argument that aren't shown with the other
two systems, as those systems are so dominated by their primary frequency
that relaxing the secular approximation doesn't seem to change much in
terms of the results. The happy side effect is that I've had to fix some
issues with dimensionality that Eric mentioned, such that it ended up being
good bug testing for FLiMESolve as well.

I'm hoping to have another pull request ready for review by the end of this
week or early next week, but I'm trying to really put everything under a
magnifying glass to the best of my ability, such that I don't leave Eric
wading through a bunch of formatting and code issues again.

On Wed, Jul 5, 2023 at 3:33 PM Simon Cross ***@***.***> wrote:

> @magnamancer <https://github.com/magnamancer> I think part of the issue
> was that you merged qutip-5.0.X into master but that shouldn't have been
> done because master now contains the latest 5.X qutip code.
>
> Sorry to be a pain, but could you open a new new PR with just your
> changes? No need to be fancy -- just create a new branch from master, copy
> across the files you want to change, create PR.
>
> The VERSION file should not be changed.
>
> How hard would it be to add tests? I remember you have tests in a notebook
> somewhere. If you can point me at the latest versions, I can think about
> how to convert them into unit tests and add them to the new PR.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/qutip/qutip/pull/2186#issuecomment-1622360050>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AV5WL5AU2Z2WLCVH7BFC2QLXOW6QNANCNFSM6AAAAAAZWI6GRQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Best Regards,
Fenton Clawson

hodgestar commented:
That's great, Fenton. Let us know when you're ready for a next round of reviews.

magnamancer commented:
Hey Simon and Eric,

- I've gone through and fixed (I hope!) the formatting and code issues due to the bad merge I did last week
- I've added in flimesolve_test.py as a test script for flimesolve
- I've tried to format everything according to PEP8, although there are likely some issues here. I've not had any formal training on PEP8 formatting (and python in general, to be honest), such that even with the help of the PEP8 style guide and Spyder's auto-formatting tool for PEP8, I'm sure there are some small fixes that need to be made
- I'm not entirely sure what Eric meant about separating out the physics from the data container, etc. I moved anything having to do with the precalculation of the Floquet states into its own function, but to kill this precalculation off entirely would severely hinder the speed of flimesolve, as an large advantage of Floquet theory comes from being able to accurately create the basis states at any time once you have the initial basis states. If this is still an issue (or if I've misinterpreted Eric's comment entirely and need to change something else in this regard), please let me know. Maybe we could have a live discussion over the issue, if that would work better.

Thanks for your help!

hodgestar commented:
Hi Fenton,

The changes look much cleaner now -- thanks! I will review when I can and hopefully Eric will have a look too.

Regarding PEP8, I recommend using an automatic tool like flake8 to check.

Ciao,
Simon

Ericgig commented on qutip/tests/solver/test_flimesolve.py:
```diff
@@ -0,0 +1,199 @@
+
+        np.testing.assert_allclose(np.real(p_ex), np.real(p_ex_ref), atol=1e-4)
+
+    def testFloquetMasterEquation3(self, kmax):
```
`kmax`  is no longer used.

Ericgig commented on qutip/tests/solver/test_flimesolve.py:
```diff
@@ -0,0 +1,199 @@
+"""
+
+
+class TestFloquet:
```
```suggestion
class TestFlimesolve:
```

Ericgig commented:
Could you send the equation for the case where `time_sense` is not `0`.

When I try it with a period of `1.0` it give extra term with very low frequencies:
```
dict_keys([0.0, 4.304872064735899e-09, 1.0157988986038049e-08, 1.1379302858450767e-08, -2.9820390185456766e-09, -4.304872064735899e-09, 5.8531169213021495e-09, 7.074430793714867e-09, -7.286911083281576e-09, -1.0157988986038049e-08, -5.8531169213021495e-09, 1.2213138724127176e-09, -1.3140028004583725e-08, -1.1379302858450767e-08, -7.074430793714867e-09, -1.2213138724127176e-09, -1.4361341876996443e-08, 2.9820390185456766e-09, 7.286911083281576e-09, 1.3140028004583725e-08, 1.4361341876996443e-08])
```

 Is it normal? You did not include any test using `time_sense`.

​

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,631 @@
+        Rate_matrix_timedep_list = [R0, *Rt_timedep_pairs[1::]]
+
+        if time_sense == 0:
+            self.solver_options["method"] = "diag"
```
This overwrite the default not the active options.
It also overwrite the users choice, it's fine to have the default differ whether the tensor is time dependent or not, but the user should choice should be respected.

magnamancer commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,631 @@
+        elif state0.type == 'oper':
+            state0 = operator_to_vector(state0)
+        else:
+            print('You need to supply a valid ket or operator')
```
Changed to raise a TypeError with the same message.

hodgestar replied:
Still needs fixing. :)

magnamancer commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,631 @@
+            self.floquet_states = None
+        super()._post_init()
+
+    def add(self, t, state):
+        if self.options["store_floquet_states"]:
+            self.floquet_states.append(state)
+        super().add(t, state)
```
This was a mistake on my part. When using flimesolve, the states are output in the computational basis, such that if the Floquet states are desired, the state needs to be transformed back into the Floquet basis using the "floquet_basis.to_floquet_basis" method. Fixing this now.

magnamancer commented:
> Could you send the equation for the case where `time_sense` is not `0`.
> 
> When I try it with a period of `1.0` it give extra term with very low frequencies:
> 
> ```
> dict_keys([0.0, 4.304872064735899e-09, 1.0157988986038049e-08, 1.1379302858450767e-08, -2.9820390185456766e-09, -4.304872064735899e-09, 5.8531169213021495e-09, 7.074430793714867e-09, -7.286911083281576e-09, -1.0157988986038049e-08, -5.8531169213021495e-09, 1.2213138724127176e-09, -1.3140028004583725e-08, -1.1379302858450767e-08, -7.074430793714867e-09, -1.2213138724127176e-09, -1.4361341876996443e-08, 2.9820390185456766e-09, 7.286911083281576e-09, 1.3140028004583725e-08, 1.4361341876996443e-08])
> ```
> 
> Is it normal? You did not include any test using `time_sense`.
> 
> ​

The relaxation of the secular approximation isn't by direct time-dependence. It was originally, but this was later changed to relax the secular approximation according to the "importance" of each term, which is dictated by a quotient. The numerator of the quotient is the product of the FFT coefficients that make up a certain element of the rate matrix super operator, and the denominator is the difference of frequencies between the two elements multiplied by the system frequency omega.

When 'timesense' is '0', the rate matrix is constructed using only the product of FFT coefficients for which the frequency arguments completely cancel. When 'timesense' isn't '0', however, is when that quotient comes into play. In a nutshell, this ends up dividing the rate at which a certain element changes the system by the frequency at which that rate rotates in the Floquet Basis. If the result of the quotient is small, the frequency is much higher than the collapse operator elements actually affect the system, such that it averages out on longer timescales. If the result of the quotient is large, then the collapse operator elements appreciably change the system on a shorter timescale than they rotate, such that their contribution to the overall matrix can't be ignored.

If I've explained this poorly, which I suspect is the case since I'm not sure how to add equations here or if that's possible, I can email you with a more thorough explanation. 

I haven't included tests for the latter case yet mostly due to my only dealing with systems heavily dictated by their primary frequency, so far. I'm working on a system now that should show different behavior for different secular approximations, but my results look slightly different from the published results such that I'm not sure I'm very confident in what I've done, yet. I'll try to update with a test for this when I can, however.

magnamancer commented:
Hey Eric,

I've been working on improving the underlying structure of FLiMESolve and
making it more efficient. I haven't done another review request yet because
I want to try to address as many things as I can at once. I've been delayed
heavily by a system I'm trying to solve for my paper about this solver, on
which Simon has offered to help such that I'll hopefully be moving forward
again soon. Additionally, there are some changes I want to make that I'm
not sure how to do which I'm going to try to address with Simon when we are
able to meet next. Most of those are in regards to structural things, e.g.
how to filter input "args" arguments to pull out the frequency of the
Hamiltonian in the same manner as other solvers in QuTiP are able to filter
arguments. I've been trying to avoid submitting review requests/questions
in the PR comments too much so as not to be a bother. I have definitely not
abandoned this PR, and plan on seeing it through fully, although I
apologize for my relative lack of communication the last few weeks. Thanks!


On Mon, Aug 21, 2023, 11:29 AM Eric Giguère ***@***.***>
wrote:

> @magnamancer <https://github.com/magnamancer> are you waiting for review,
> still working on the PR or abandoned it?
>
> There are still issues I raised here and in magnamancer#4
> <https://github.com/magnamancer/qutip/pull/4>, that you have not been
> addressed and I am waiting for them before the next round of review.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/qutip/qutip/pull/2186#issuecomment-1686550860>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AV5WL5GN3R6GPKYKYNOWSGDXWN5GDANCNFSM6AAAAAAZWI6GRQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

hodgestar commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,646 @@
+
+    total_R_tensor = {}
+    for cdx, c_op in enumerate(c_ops):
+        '''
+        Transforming the lowering operator into the Floquet Basis
+            and taking the FFT
+        '''
```
```suggestion
        # Transforming the lowering operator into the Floquet Basis
        # and taking the FFT:
```

hodgestar commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,646 @@
+            np.fft.fft(c_op_Floquet_basis, axis=0) / len(tlist)
+        )
+
+        '''
+        Next step is to form arrays of the rate products (rate_products)
+            and the frequency at which these components rotate (delta_array),
+            then dividing the arrays to form the rate quotient array.
+        
+        In quotient array, setting any undefined value, i.e. those for which
+            rate_products = 0, equal to time_sense+1 so that the mask
+            arrays mask out these values.
+        '''
```
```suggestion
        # Next step is to form arrays of the rate products (rate_products)
        # and the frequency at which these components rotate (delta_array),
        # then dividing the arrays to form the rate quotient array.
        
        # In quotient array, setting any undefined value, i.e. those for which
        # rate_products = 0, equal to time_sense+1 so that the mask
        # arrays mask out these values.
```

hodgestar commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,646 @@
+            something Rate Matrix something something linear Operator.
+    '''
+    Hdim = len(floquet_basis.e_quasi)  # Dimensionality of the Hamiltonian
+    args_key = list(args.values())[0]  # Frequency dependence of Hamiltonian
```
Perhaps we can pass in `T` instead of `args` and then set the frequency for rescaled to `1 / T` or similar?

hodgestar replied:
P.S. `T` is also missing from the docstrings.

hodgestar commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,646 @@
+        '''
+        # Masking any frequency greater than Nyquist Frequency
+        nyquist_mask = np.ma.getmaskarray(np.ma.masked_greater(
+            delta_array, np.pi/(tlist[1]-tlist[0])))
```
Could `tlist[1] - tlist[0]` be just `dt`?

hodgestar commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,646 @@
+        quotient_array = np.divide(
+            args_key*delta_array,
+            abs(rate_products),
+            out=np.full(np.shape(delta_array), (time_sense+1), dtype=complex),
```
Let's perhaps add a comment here explaining that the `out` value is set to provide defaults outside of the quotient mask range where the rate_products are zero.

hodgestar commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,646 @@
+        # Grouping valid_indices by their frequency values
+        delta_dict = {}
+        for idx, indices in enumerate(valid_indices):
+            try:
+                delta_dict[delta_vals_truncd[idx]].append(indices)
+            except KeyError:
+                delta_dict[delta_vals_truncd[idx]] = [indices]
```
```suggestion
            delta_dict.setdefault(delta_vals_truncd[idx], []).append(indices)
```
Alternative you can use `defaultdict` from collections and just write `delta_dict = defaultdict(list)` when defining `delta_dict`. Then just do `delta_dict[delta_vals_truncd[idx]].append(indices)`.

hodgestar commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,646 @@
+                delta_dict[delta_vals_truncd[idx]] = [indices]
+
+        # For each valid frequency, form a slice of total_R_tensor
+        for key in delta_dict.keys():
```
```suggestion
        for key in delta_dict:
```

hodgestar replied:
My guess here is that this is slow because we're combining loops and big masked arrays. Perhaps we can do everything with the masked array we originally constructed?

hodgestar commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,646 @@
+    rho0 / psi0 : :class:`qutip.Qobj`
+        Initial density matrix or state vector (ket).
+
+    tlist:*list* / *array*
```
```suggestion
    tlist : list or array
```

hodgestar commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,631 @@
+            self.floquet_basis.T,
+            c_ops,
+            c_op_rates,
+            Hargs,
```
We can just remove `Hargs` and use `T` instead.

hodgestar commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,646 @@
+
+        # Rate_Qobj_list = [Qobj(
+        #     RateMat, type="super", superrep="super", copy=False
+        Rate_Qobj_list = [Qobj(
```
Here you can use `.to("csr")` to make these sparse (which will perhaps be much faster if they are sparse).

hodgestar commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,646 @@
+        #     rate_matrix.dims =  floquet_basis.U(0).dims
+        R0 = Rate_Qobj_list[0]
+
+        Rt_timedep_pairs = [list(
```
You can drop the `list(` here because you have a list comprehension inside so it already creates a list. :)

hodgestar commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,646 @@
+                self.Nt = 2**4
+        self.time_sense = time_sense
+
+        RateDic = _floquet_rate_matrix(
```
Let's create a method named `_create_rhs` or similar on the FLiMESolver class and calculate the RHS in one function, rather than making some numpy arrays and then converting them to `QObj` outside.

Inside the `_create_rhs` method we can also add the timing for the creation of the RHS to the stats.

See https://github.com/qutip/qutip/blob/2dc5f3f0a6ad2175e6dc015649989125401fc3bf/qutip/solver/heom/bofin_solvers.py#L661 and then https://github.com/qutip/qutip/blob/2dc5f3f0a6ad2175e6dc015649989125401fc3bf/qutip/solver/heom/bofin_solvers.py#L684.

hodgestar reviewed (changes requested):
Left some comments, but looking good, I think!

hodgestar commented:
Let's also try add a test in the region where FLiMESolve and the older FloquetSolver agree.

magnamancer commented:
Okay, so I've gone over Eric's branch of my code and more or less implemented all of his changes. Everything seems to be in working order, so I'm requesting another review.

magnamancer commented:
Hey @Ericgig, I'm commenting here as you requested last night, to get a quick review of FLiMESolve again. Thanks!

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,654 @@
+        List of 2**n times distributed evenly over one period of the
+            Hamiltonian
+
+    taulist: np.ndarray
+        List of number_of_periods*2**n times distributed evenly over the
+            entire duration of Hamiltonian driving
```
No longer a parameter.

Ericgig reviewed (commented):
More polishing is needed.
You have unused variables, some docstring are not matching code, it does not fit qutip conventions in many places, the `step` method seems broken, etc.

For now, could you fix the conflict so we can run tests.

magnamancer commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,654 @@
+        self,
+        floquet_basis,
+        c_ops_and_rates,
+        Hargs,
```
It was originally used to import the frequency of the Hamiltonian. I think my thought in keeping it was to allow for generalization to MMFT later, but I'll take it out now!

magnamancer commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,654 @@
+        time_sense=0,
+        options=None
+    ):
+        if time_sense == 0:
+            self.solver_options["method"] = "diag"
```
I'm honestly unsure of how to properly set this without overwriting the default. I'll try to figure it out!

magnamancer replied:
I changed the method to None and added a bit in the documentation describing what happens.

Ericgig replied:
Where did you document this?
It still change the class instance and will not be `None` after the first call.

magnamancer replied:
I added a sentence or two to the guide, but realized by "documentation," you actually meant the info under the definitions. Added sentences in the parameter documentation for FLiMESolve, FLiMESolver to explain time_sense in a nutshell, as well as which solver is used and when. Let me know if it's adequate, please!

magnamancer commented:
@Ericgig Hey Eric, I fixed the test script issues last week, but wasn't sure if you wanted me to ping you again. I'm pinging you again now, but if this isn't what you would like, please let me know!

That is, I'm _pretty sure_ everything should be working now. The test_flimesolve.py script, flimesolve (with your changes made - without the fixing of the step method yet), and correlation.py (where I added options for FLiMESolve to be used) should all be working now. Let me know if there's anything else that needs fixed. Thank you!

magnamancer commented:
@Ericgig Hey again Eric, sorry for the double ping. Small issue where I forgot to turn the returned Floquet states (when desired) as quantum objects instead of just numpy arrays. This caused one of the example scripts to mess up again. I fixed this and it looks like the script runs on my end.

Ericgig reviewed (commented):
Getting close, there are a few places where the documentation is not up to date with the latest changes and some style issues, but it's getting there.

It would be great to have reference to a paper which describe the method and show the equations used if you have one.

You can add to flimesolve to import in `qutip/solver/__init__.py` so it can be included in qutip namespace.

I would prefer if you did not convert to numpy arrays but worked with qutip data as much as possible. This would make JAX, cupy support easier to add down the line, but this can be fixed later in another PR.

magnamancer commented on doc/guide/scripts/floquet_ex4.py:
```diff
@@ -0,0 +1,54 @@
+    args=args, options={"store_floquet_states": True}
+)
+
+# calculate expectation values in the computational basis
+p_ex = np.zeros(tlist.shape, dtype=np.complex128)
+for idx, t in enumerate(tlist):
+    f_coeff_t = output.floquet_states[idx]
+    psi_t = output.floquet_basis.from_floquet_basis(f_coeff_t, t)
+    # Alternatively
+    psi_t = output.states[idx]
+    p_ex[idx] = qutip.expect(qutip.num(2), psi_t)
```
This was specifically to match the original examples as much as possible. I can change it if you think that's better, though!

magnamancer commented on doc/guide/dynamics/dynamics-floquet.rst:
```diff
@@ -335,3 +335,46 @@ expectation values using:
     :include-source: false
     :nofigs:
+
+The Floquet-Lindblad master equation in QuTiP
+---------------------------------------------
```
Fixed the import issue. Sent you a message about the other part of this comment

Ericgig commented:
Did you push all your commits? Some of the comment I made that you marked as resolved are not.

magnamancer commented:
@Ericgig Pushed all the changes, this time.

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,655 @@
+    time_sense=0,
+    options={},
+):
+    """
+    Parameters
```
Adding a description of the solver here would be nice.

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,655 @@
+
+
+class FLiMESolver(MESolver):
+    """
+     Solver for the Floquet-Markov master equation.
+
```
Is this still right? Shouldn't this be Floquet-Lindblad master equations?

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,648 @@
+             Hamiltonian
+
+    c_ops : list of :class:`qutip.Qobj`.
+        List of lists of [collapse operator, collapse operator rate] pairs
```
The rate is now merged in the operator.

magnamancer replied:
Writing a comment specifically to note that I fixed this, this time. If it doesn't show up in the push, it's cause I messed up the merge again.

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,648 @@
+        pull Nt from tlist.
+
+    c_ops : list of :class:`qutip.Qobj`.
+        List of lists of [collapse operator,collapse operator rate] pairs
```
This is not fixed!

Ericgig reviewed (commented):
- There are no tests with `time_sense` not `0`.
- These are no tests with correlation using `flimesolve`.

magnamancer commented:
@Ericgig I fixed the typos you mentioned and added a test in for nonzero timesense values and for the correlation functions. I also fixed the typo you mentioned and added a short descriptor for FLiMESolve above the "parameters" in documentation.

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,666 @@
+
+    time_sense : float
+        Value of the secular approximation to use when constructing the rate
+        matrix R(t).Default value of zero uses the fully time-independent/most
```
```suggestion
        matrix R(t). Default value of zero uses the fully time-independent/most
```

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,666 @@
+    time_sense : float
+        Value of the secular approximation to use when constructing the rate
+        matrix R(t).Default value of zero uses the fully time-independent/most
+        strict secular approximation, and will utilize the "diag" solver method.
+        Values greater than zero have time dependence, and will subsequently
+        use the "Adams" method for the ODE solver.
```
The way it written it seems that the user input of `method` will be ignored in favor of these.
It only the default that is changing.
```suggestion
        strict secular approximation, values greater than zero have time 
        dependence. The default integration method change depending
        on this value, "diag" for `0`, "adams" otherwise.
```

Ericgig commented on qutip/tests/solver/test_flimesolve.py:
```diff
@@ -0,0 +1,293 @@
+        Test Floquet-Lindblad Master Equation with correlation functions.
+
+        """
+        E1mag = 2 * np.pi * 0.072992700729927
+        E1 = E1mag
+        d = 1
+
+        Om1 = d * E1
```
```suggestion
        Om1 = 2 * np.pi * 0.072992700729927
```

Ericgig commented on qutip/tests/solver/test_flimesolve.py:
```diff
@@ -0,0 +1,293 @@
+        dt = timef / Nt
+        tlist = np.linspace(0, timef - dt, Nt)
+
+        H_atom = ((wres - wlas) / 2) * Qobj([[-1, 0], [0, 1]])
+        Hf1 = -(1 / 2) * Qobj([[0, Om1], [np.conj(Om1), 0]])
+
+        H = [H_atom + Hf1]
```
`wres == wlas`, `Om1` is real:
```suggestion
        H = -(Om1 / 2) * qt.sigmax()
```

Ericgig commented on qutip/tests/solver/test_flimesolve.py:
```diff
@@ -0,0 +1,293 @@
+        Hargs = {"l": (wlas)}
+        Gamma = 2 * np.pi * 0.00025
+
+        Nt = 1600
+        timef = 10 * T
```
This is a test, we want a quick computation:
```suggestion
        Nt = 20
        timef = 2 * T
```

Ericgig commented on qutip/tests/solver/test_flimesolve.py:
```diff
@@ -0,0 +1,293 @@
+        wres = 2 * np.pi * 280
+
+        T = 2 * np.pi / wlas
+        Hargs = {"l": (wlas)}
```
```suggestion
```

Ericgig commented on qutip/tests/solver/test_flimesolve.py:
```diff
@@ -0,0 +1,293 @@
+
+        Om1 = d * E1
+
+        wlas = 2 * np.pi * 280
+        wres = 2 * np.pi * 280
```
`Hargs` is never used.
`H_atom` is `0` and can be removed.
So they end up not being used.
```suggestion
```

magnamancer commented:
@Ericgig Hey Eric, added your changes to the test, fixed the documentation error as well. I also added a steady state solver, would you want me to make a test for that?

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,679 @@
+            steady state of the system. Note: This only uses the full secular
+            approximation.
+        """
+        monomat = self.RateDic[0].full()
+        val, vec = np.linalg.eig(monomat)
```
Qutip wrap numpy/scipy eigen sovler, please do not use it directly, but use `Qobj.eigenstates` or other qutip function. In general, any place `Qobj.full` is used means that we will have to fix it later to allow it to run with other back-ends (jax, cupy, ...).

I would prefer you not adding new features at this point and add them in another PR. Otherwise there is always more to review and pushing the eventual merging back.

magnamancer replied:
Sounds good. I'll roll it back.

Ericgig reviewed (commented):
I discussed with @hodgestar and we want citation(s) to paper(s) where the equations are explained added before merging.

magnamancer commented:
Hey @Ericgig and @hodgestar, I'm tagging you with an ArXiv link to the paper. I hope everything looks okay!

link: https://arxiv.org/abs/2410.18046

Ericgig reviewed (commented):
Could you merge the current master to trigger the tests.

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,669 @@
+
+        self.options = options
+
+        self.Nt = self.options["Nt"]
```
Now updating `Nt` will update the equation, but not `time_sense`.
I believe that both should be treated the same way.
Also using the setter to compute the rhs is strange, it make the code harder to read as the reader will not expect that `self.Nt = self.options["Nt"]` compute the floquet tensor, the rhs matrix, initiate the integrator and reset the state metadata...

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,691 @@
+    else:
+        floquet_basis = FloquetBasis(H, T, args, precompute=None)
+
+        if Nt != None:
```
```suggestion
        if Nt is not None:
```

Ericgig commented on qutip/solver/flimesolve.py:
```diff
@@ -0,0 +1,691 @@
+    rho0,
+    tlist,
+    T,
+    Nt=None,
+    c_ops=None,
+    *,
+    e_ops=None,
+    args=None,
+    time_sense=0,
```
In the class interface, `time_sense` is a positional arguments, here it is keyword only.
But the class interface does not have `Nt`... 
The tests always jump over `Nt` to input the collapse operators.
How about having `Nt` keyword only or having it in options like it is for the class interface.

Ericgig commented on qutip/tests/solver/test_flimesolve.py:
```diff
@@ -0,0 +1,312 @@
+            options={"store_floquet_states": True},
+            time_sense=t_sensitivity,
+            e_ops=num(2),
+            Nt=int(2**8 / 50),
```
Why this equation?  Why not just set the number as an integer?

Ericgig commented on qutip/tests/solver/test_flimesolve.py:
```diff
@@ -0,0 +1,312 @@
+            c_ops=[c_op * np.sqrt(gamma1)],
+            e_ops=e_ops,
+            args=args,
+            Nt=50,
```
You added a none default value for `Nt` in every test.
Are test failing with the default value or it is there just to ensure the parameter is tested.

Ericgig commented on qutip/solver/correlation.py:
```diff
@@ -414,16 +504,30 @@ def coherence_function_g2(H, state0, taulist, c_ops, a_op, solver="me",
+            c_ops,
+            time_sense=kwargs["time_sense"] if "time_sense" in kwargs else 0,
+            options=options,
+            Nt=kwargs["Nt"] if "Nt" in kwargs else 2**4,
```
```suggestion
            Nt=kwargs.get("Nt", 16),
```

Ericgig commented on qutip/tests/solver/test_flimesolve.py:
```diff
@@ -0,0 +1,285 @@
+        ).expect[0]
+        # Compare with mesolve
+        p_ex_ref = mesolve(
+            H, psi0, tlist, [np.sqrt(gamma1) * c_op], e_ops, args
```
```suggestion
            H, psi0, tlist, [np.sqrt(gamma1) * c_op], e_ops=e_ops, args
```

Ericgig commented on qutip/tests/solver/test_flimesolve.py:
```diff
@@ -0,0 +1,285 @@
+        ).expect[0]
+        # Compare with mesolve
+        p_ex_ref = mesolve(
+            H, psi0, tlist, [np.sqrt(gamma1) * c_op], e_ops=e_ops, args
```
```suggestion
            H, psi0, tlist, [np.sqrt(gamma1) * c_op], e_ops=e_ops, args=args
```
